File: 20-valid-parentheses/valid-parentheses.py
class Solution(object):
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        
        check = {
            ')' : '(',
            '}' : '{',
            ']' : '['
        }

        ans = []

        for i in s:
            
            if i not in check:
                ans.append(i)

            elif ans and check[i] == ans[-1]:
                ans.pop()
            else:
                return False

        if not ans:
            return True
        else:
            return False

File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        
        HashMap<Character, Character> check = new HashMap<>();

        check.put(')', '(');
        check.put('}', '{');
        check.put(']', '[');

        Stack<Character> ans = new Stack<>();

        for (char i : s.toCharArray()){
            if (!check.containsKey(i)){
                ans.push(i);
            }
            else if (!ans.isEmpty() && check.get(i) == ans.peek()){
                ans.pop();
            }
            else {
                return false;
            }
        }
        return ans.isEmpty();
    }
}